src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -0,0 +1,382 @@
+            logger.trace("Searching memtable index on expression '{}'...", expression);
+
+        switch (expression.getOp())
+        {
+            case EQ:
+            case CONTAINS_KEY:
+            case CONTAINS_VALUE:
+                return exactMatch(expression);
+            case RANGE:
+                return rangeMatch(expression, keyRange);
+            default:
+                throw new IllegalArgumentException("Unsupported expression: " + expression);
+        }
+    }
+
+    /**
+     * Returns an {@link Iterator} over the entire dataset contained in the trie. This is used
+     * when the index is flushed to disk.
+     *
+     * @return the iterator containing the trie data
+     */
+    public Iterator<Pair<ByteComparable, PrimaryKeys>> iterator()
+    {
+        Iterator<Map.Entry<ByteComparable, PrimaryKeys>> iterator = data.entrySet().iterator();
+        return new Iterator<Pair<ByteComparable, PrimaryKeys>>()
+        {
+            @Override
+            public boolean hasNext()
+            {
+                return iterator.hasNext();
+            }
+
+            @Override
+            public Pair<ByteComparable, PrimaryKeys> next()
+            {
+                Map.Entry<ByteComparable, PrimaryKeys> entry = iterator.next();
+                return Pair.create(decode(entry.getKey()), entry.getValue());
+            }
+        };
+    }
+
+    public ByteBuffer getMinTerm()
+    {
+        return minTerm;
+    }
+
+    public ByteBuffer getMaxTerm()
+    {
+        return maxTerm;
+    }
+
+    private void setMinMaxTerm(ByteBuffer term)
+    {
+        assert term != null;
+
+        minTerm = TypeUtil.min(term, minTerm, indexContext.getValidator());
+        maxTerm = TypeUtil.max(term, maxTerm, indexContext.getValidator());
+    }
+
+    private ByteComparable encode(ByteBuffer input)
+    {
+        return isLiteral ? version -> append(ByteSource.of(input, version), ByteSource.TERMINATOR)
+                         : version -> TypeUtil.asComparableBytes(input, validator, version);
+    }
+
+    private ByteComparable decode(ByteComparable term)
+    {
+        return isLiteral ? version -> ByteSourceInverse.unescape(ByteSource.peekable(term.asComparableBytes(version)))
+                         : term;
+
+    }
+
+    private ByteSource append(ByteSource src, int lastByte)
+    {
+        return new ByteSource()
+        {
+            boolean done = false;
+
+            @Override
+            public int next()
+            {
+                if (done)
+                    return END_OF_STREAM;
+                int n = src.next();
+                if (n != END_OF_STREAM)
+                    return n;
+
+                done = true;
+                return lastByte;
+            }
+        };
+    }
+
+    private RangeIterator exactMatch(Expression expression)
+    {
+        final ByteComparable prefix = expression.lower == null ? ByteComparable.EMPTY : encode(expression.lower.value.encoded);
```

Comment:
Is this a prefix, or a path w/ a terminator? (In other words, this isn't a prefix search, but a search for an exact term?)

Reply:
It's a path with a terminator. I'll rename this to reflect this.